core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java

```diff
@@ -140,65 +140,90 @@ protected IndexResponse newResponseInstance() {
     }
 
     @Override
-    protected WriteResult<IndexResponse> onPrimaryShard(IndexRequest request, IndexShard indexShard) throws Exception {
-        return executeIndexRequestOnPrimary(request, indexShard, mappingUpdatedAction);
+    protected WritePrimaryResult shardOperationOnPrimary(IndexRequest request, IndexShard primary) throws Exception {
+        final Engine.IndexResult indexResult = executeIndexRequestOnPrimary(request, primary, mappingUpdatedAction);
+        final IndexResponse response = indexResult.hasFailure() ? null :
+                new IndexResponse(primary.shardId(), request.type(), request.id(), indexResult.getVersion(),
+                        indexResult.isCreated());
+        return new WritePrimaryResult(request, response, indexResult.getLocation(), indexResult.getFailure(), primary);
     }
 
     @Override
-    protected Location onReplicaShard(IndexRequest request, IndexShard indexShard) {
-        return executeIndexRequestOnReplica(request, indexShard).getTranslogLocation();
+    protected WriteReplicaResult shardOperationOnReplica(IndexRequest request, IndexShard replica) throws Exception {
+        final Engine.IndexResult indexResult = executeIndexRequestOnReplica(request, replica);
+        return new WriteReplicaResult(request, indexResult.getLocation(), indexResult.getFailure(), replica);
     }
 
     /**
      * Execute the given {@link IndexRequest} on a replica shard, throwing a
      * {@link RetryOnReplicaException} if the operation needs to be re-tried.
      */
-    public static Engine.Index executeIndexRequestOnReplica(IndexRequest request, IndexShard indexShard) {
-        final ShardId shardId = indexShard.shardId();
+    public static Engine.IndexResult executeIndexRequestOnReplica(IndexRequest request, IndexShard replica) {
+        final ShardId shardId = replica.shardId();
         SourceToParse sourceToParse = SourceToParse.source(SourceToParse.Origin.REPLICA, shardId.getIndexName(), request.type(), request.id(), request.source())
                 .routing(request.routing()).parent(request.parent()).timestamp(request.timestamp()).ttl(request.ttl());
 
-        final Engine.Index operation = indexShard.prepareIndexOnReplica(sourceToParse, request.version(), request.versionType(), request.getAutoGeneratedTimestamp(), request.isRetry());
+        final Engine.Index operation;
+        try {
+            operation = replica.prepareIndexOnReplica(sourceToParse, request.version(), request.versionType(), request.getAutoGeneratedTimestamp(), request.isRetry());
+        } catch (MapperParsingException | IllegalArgumentException e) {
```

Comment:
It should never happen that we get this specific exception on the replica if we did not get it on the primary before as it would mean that two shards have a different type for a given field even though types cannot be changed once a field is added.